source/Coop/Mod/DebugUtil/MoveTo.cs

```diff
@@ -0,0 +1,283 @@
+
+                case MoveToState.Playback:
+                    MoveToPlayback -= OnEventPlayback;
+                    MoveToRecording -= OnEventRecording;
+                    state = MoveToState.Stop;
+
+                    VerifyEvents();
+
+                    RecordingEventList = null;
+                    PlaybackEventList = null;
+                    PlaybackMainPartyList = null;
+                    return $"Playback file '{current}' stopped.";
+
+                default:
+                    return null;
+            }
+        }
+
+        private static void OnEventRecording(EntityId entityId, MobileParty party, MovementData movement)
+        {
+            RecordingEventList.Add(new MoveToEvent()
+            {
+                time = CampaignTime.Now,
+                entityId = entityId,
+                party = party,
+                movement = movement
+            });
+            if (party.IsPlayerControlled())
+                Logger.Warn("[MOVETO] Yet one player's moving recorded.");
+        }
+
+        public static void OnEventPlayback()
+        {
+            var now = CampaignTime.Now;
+
+            if (firstTick <= now)
+            {
+                MoveToRecording += OnEventRecording;
+                firstTick = CampaignTime.Never;
+            }
+
+            var moveTo = PlaybackMainPartyList.FirstOrDefault(q => !q.passed && q.time <= now);
+            if (moveTo != null)
+            {
+                if (CoopServer.Instance.Persistence.Room.Entities.
+                    FirstOrDefault(q => q.Id == moveTo.entityId) is MobilePartyEntityServer entity)
+                    entity.State.Movement = new MovementState()
+                    {
+                        DefaultBehavior = moveTo.movement.DefaultBehaviour,
+                        Position = moveTo.movement.TargetPosition,
+                        SettlementIndex = moveTo.movement.TargetSettlement != null ?
+                            moveTo.movement.TargetSettlement.Id : MovementState.InvalidIndex,
+                        TargetPartyIndex = moveTo.movement.TargetParty != null ?
+                            moveTo.movement.TargetParty.Id : MovementState.InvalidIndex
+                    };
+
+                moveTo.passed = true;
+                Logger.Warn("[MOVETO] Moving to new position.");
+            }
+
+            if (now >= lastTick)
+            {
+                Stop(); // TODO: send message in debug console instead to send on game screen
+                Logger.Warn("[MOVETO] Playback has finished.");
+            }
+        }
+
+        private static void VerifyEvents()
+        {
+            var difftime = new List<long?>();
+
+            foreach (var play in PlaybackEventList)
+            {
+                var rec = RecordingEventList.FirstOrDefault(q => !q.passed && q.Equals(play));
+                if (rec != null)
+                {
+                    difftime.Add((long)(rec.time - play.time).ToMilliseconds);
```

Comment:
I think there's an issue here with what is being measured. `play.time` is the campaign time when the client _received_ the event during the first recording. During playback, the event was was processed server side once the campaign time was higher than that time. `rec.time` is again the time the client received the event during playback. Meaning what you're measuring here is the ingame time difference for:

- detecting the state change on the server side (1 railgun update call on server)
- sending the new server to the clients (no additional update call)
- receiving the new state on the client (1 network receive call)
- processing the new state on the client (1 railgun update call on client)

So thats 1 frame on the server and 1 frame on the client, that is the absolute minimum you can measure using this approach. Is that intended?

Reply:
@araex It's not problem for desynchronization analyzing. We just search abnormal times. **But** it will be problem if that delay in 2 ticks affects on how local game instance reacts in contrast server's instance.